### PR TITLE
feat(plonky2): implemented less-than using bitwise operations

### PR DESCRIPTION
**Pull request by nickysn**
The previous less-than implementation used division, which is less optimal and
caused problems with comparisons with 0 (which caused division by 0).

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs [hunk 1]
```diff
@@ -603,10 +603,75 @@ impl Builder {
...
+                            P2Type::Field => FIELD_BIT_SIZE,
+                            _ => {
+                                let message = format!(
+                                    "less than op invoked on arguments of type {:?}",
+                                    bit_size_a
+                                );
+                                return Err(Plonky2GenError::ICE { message });
+                            }
+                        })
+                        .unwrap();
```
Comment (smanilov) [inline compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs:617]: Could we please have this outlined in a function named `self.get_integer_bitsize(type_of_a)`?

File: compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs [hunk 2]
```diff
@@ -603,10 +603,75 @@ impl Builder {
                         let (bit_size_b, target_b) = self.get_integer(rhs)?;
                         assert!(bit_size_a == bit_size_b);
 
-                        let div = add_div_mod(&mut self.asm_writer, target_a, target_b).0;
-                        let zero = self.asm_writer.zero();
-                        let target = self.asm_writer.is_equal(div, zero);
-                        Ok(P2Value::make_boolean(target))
+                        let bit_size = usize::try_from(match bit_size_a {
```
Comment (smanilov) [inline compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs:606]: Looks like `bit_size_a` is actually `type_of_a`. Perhaps the function was changed at some point but the use site was not updated.

Please, rename.

File: compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs [hunk 3]
```diff
@@ -603,10 +603,75 @@ impl Builder {
...
+                        split_b.reverse();
+
+                        // generate:
+                        //   (!a[0] and b[0]) or
+                        //   ((a[0] == b[0]) and ((!a[1] and b[1]))) or
+                        //   ((a[0] == b[0]) and (a[1] == b[1]) and (!a[2] and b[2])) or
+                        //   ((a[0] == b[0]) and (a[1] == b[1]) and (a[2] == b[2]) and (!a[3] and b[3])) or ...
+                        //   ...
+                        //   ((a[0] == b[0]) and ... and (a[i-1] == b[i-1]) and (!a[i] and b[i])) or ...
+                        //   ...
```
Comment (smanilov) [inline compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs:632]: I love the comment!

**Review by smanilov — CHANGES_REQUESTED** [review]
Nice!

Although it looks quite good to me, I will only stop worrying once we have a test that does more than a few comparisons and asserts their correct outcome.

Could you please add such a test? I think it should be okay if it's a single test, but it should perform many comparisons, so perhaps take two arrays as inputs and compare them pairwise. Something like:

```
fn beefy_test(x: pub [u32; 16], y: [u32; 16]) {
  for i in 0..7 {
    assert(x[i] < y[i]);
  }
  for i in 8..15 {
    assert((x[i] < y[i]) == false);
  }
}
```

(might not be correct Noir syntax, but you get the idea)

**Review by Aristotelis2002 — APPROVED** [review]
LGTM and I don't have anything more to add. Please, address Stan's comments.

File: compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs (hunk 2 above)
Comment (nickysn) [reply compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs]: Done: https://github.com/blocksense-network/noir/pull/60/commits/7e8fdb5aac2698d6c381d78f95e709cde00a50d5

**Comment by nickysn** [comment]
> Nice!
> 
> Although it looks quite good to me, I will only stop worrying once we have a test that does more than a few comparisons and asserts their correct outcome.
> 
> Could you please add such a test? I think it should be okay if it's a single test, but it should perform many comparisons, so perhaps take two arrays as inputs and compare them pairwise. Something like:
> 
> ```
> fn beefy_test(x: pub [u32; 16], y: [u32; 16]) {
>   for i in 0..7 {
>     assert(x[i] < y[i]);
>   }
>   for i in 8..15 {
>     assert((x[i] < y[i]) == false);
>   }
> }
> ```
> 
> (might not be correct Noir syntax, but you get the idea)

Ok, I added such a test: https://github.com/blocksense-network/noir/pull/60/commits/24f8ff37c70f299cc5ea68a438497ba9627f0bef

File: test_programs/plonky2_prove_success/less_than_4/Prover.toml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+x = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9,10,11,12,13,14,15]
+y = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9,10,11,12,13,14,15]
```
Comment (smanilov) [inline test_programs/plonky2_prove_success/less_than_4/Prover.toml:2]: These are good, could we just have some larger numbers as well?

So maybe [0, 1, 2, 13, 457, ..., 1000000001], perhaps use an RNG.

**Review by smanilov — APPROVED** [review]
Looks good, just mix up the inputs for the new tests, please (comment in-line).

File: test_programs/plonky2_prove_success/less_than_4/Prover.toml (hunk 4 above)
Comment (nickysn) [reply test_programs/plonky2_prove_success/less_than_4/Prover.toml]: Ok, I added some more bit patterns (with larger numbers) to the test: https://github.com/blocksense-network/noir/pull/60/commits/925d15d6e9475e9ce59aa3c8889d5214e38133e2

Comment (smanilov) [reply test_programs/plonky2_prove_success/less_than_4/Prover.toml]: Looks great! Merging.

**Pull request merged**